block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -0,0 +1,326 @@
+import coop.rchain.blockstorage.util.BlockMessageUtil.parentHashes
+import coop.rchain.blockstorage.util.TopologicalSortUtil
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.shared.{Log, LogSource}
+
+import scala.collection.immutable.HashSet
+import scala.util.{Failure, Success, Try}
+
+final class BlockDagFileStorage[F[_]: Monad: Concurrent: Sync: Log] private (
+    lock: Semaphore[F],
+    latestMessagesRef: Ref[F, Map[Validator, BlockHash]],
+    childMapRef: Ref[F, Map[BlockHash, Set[BlockHash]]],
+    dataLookupRef: Ref[F, Map[BlockHash, BlockMetadata]],
+    topoSortRef: Ref[F, Vector[Vector[BlockHash]]],
+    latestMessagesLogOsRef: Ref[F, OutputStream],
+    latestMessagesLogSizeRef: Ref[F, Int],
+    latestMessagesCrcRef: Ref[F, CRC32],
+    latestMessagesDataFilePath: Path,
+    latestMessagesCrcFilePath: Path,
+    latestMessagesLogMaxSizeFactor: Int
+) extends BlockDagStorage[F] {
+  private final case class FileDagRepresentation(
+      latestMessagesMap: Map[Validator, BlockHash],
+      childMap: Map[BlockHash, Set[BlockHash]],
+      dataLookup: Map[BlockHash, BlockMetadata],
+      topoSortVector: Vector[Vector[BlockHash]]
+  ) extends BlockDagRepresentation[F] {
+    def children(blockHash: BlockHash): F[Option[Set[BlockHash]]] =
+      childMap.get(blockHash).pure[F]
+    def lookup(blockHash: BlockHash): F[Option[BlockMetadata]] =
+      dataLookup.get(blockHash).pure[F]
+    def contains(blockHash: BlockHash): F[Boolean] =
+      dataLookup.contains(blockHash).pure[F]
+    def topoSort(startBlockNumber: Long): F[Vector[Vector[BlockHash]]] =
+      topoSortVector.drop(startBlockNumber.toInt).pure[F]
+    def topoSortTail(tailLength: Long): F[Vector[Vector[BlockHash]]] =
+      topoSortVector.takeRight(tailLength.toInt).pure[F]
+    def deriveOrdering(startBlockNumber: Long): F[Ordering[BlockMetadata]] =
+      topoSort(startBlockNumber).map { topologicalSorting =>
+        val order = topologicalSorting.flatten.zipWithIndex.toMap
+        Ordering.by(b => order(b.blockHash))
+      }
+    def latestMessageHash(validator: Validator): F[Option[BlockHash]] =
+      latestMessagesMap.get(validator).pure[F]
+    def latestMessage(validator: Validator): F[Option[BlockMetadata]] =
+      latestMessagesMap.get(validator).flatTraverse(lookup)
+    def latestMessageHashes: F[Map[Validator, BlockHash]] =
+      latestMessagesMap.pure[F]
+    def latestMessages: F[Map[Validator, BlockMetadata]] =
+      latestMessagesMap.toList
+        .traverse {
+          case (validator, hash) => lookup(hash).map(validator -> _.get)
+        }
+        .map(_.toMap)
+  }
+
+  private def updateLatestMessagesFile(validator: Validator, blockHash: BlockHash): F[Unit] =
+    for {
+      latestMessagesLogOs <- latestMessagesLogOsRef.get
+      latestMessagesCrc   <- latestMessagesCrcRef.get
+      toAppend            = validator.concat(blockHash).toByteArray
+      _                   = latestMessagesCrc.update(toAppend)
```

Comment:
these operations having arbitrary side-effects look pretty bad. 

We should refactor these to be in sync with the rest of our code

Reply:
True. I will create an immutable wrapper for CRC32.